src/overlap.c

```diff
@@ -0,0 +1,529 @@
+                                         checker->size)) {
+    overlap_checker_free(checker);
+    return NULL;
+  }
+
+  /* Use the perm to sort the cell. */
+  /* The sorted cell is saved for as long as the OverlapChecker lives. */
+  permute_double_3(checker->pos_sorted,
+                   cell->position,
+                   checker->perm_temp,
+                   cell->size);
+
+  permute_int(checker->types_sorted,
+              cell->types,
+              checker->perm_temp,
+              cell->size);
+
+  return checker;
+}
+
+static int argsort_by_lattice_point_distance(int * perm,
+                                             SPGCONST double lattice[3][3],
+                                             SPGCONST double (* positions)[3],
+                                             const int * types,
+                                             double * distance_temp,
+                                             void *argsort_work,
+                                             const int size)
+{
+  double diff[3];
+  int i, k;
+  double x;
+
+  /* Fill distance_temp with distances. */
+  for (i = 0; i < size; i++) {
+    /* Fractional vector to nearest lattice point. */
+    for (k = 0; k < 3; k++) {
+      x = positions[i][k];
+      diff[k] = x - mat_Nint(x);
+    }
+
+    /* Squared distance to lattice point. */
+    mat_multiply_matrix_vector_d3(diff, lattice, diff);
+    distance_temp[i] = mat_norm_squared_d3(diff);
+  }
+
+  return perm_argsort(perm,
+                      types,
+                      distance_temp,
+                      argsort_work,
+                      size);
+}
+
+/* Tests if an operator COULD be a symmetry of the cell, */
+/* without the cost of sorting the rotated positions. */
+/* It only inspects a few atoms. */
+/* 0:  Not a symmetry.   1. Possible symmetry. */
+static int check_possible_overlap(OverlapChecker *checker,
+                                  const double test_trans[3],
+                                  SPGCONST int rot[3][3],
+                                  const double symprec)
+{
+  double pos_rot[3];
+  int i, i_test, k, max_search_num, search_num;
+  int type_rot, is_found;
+
+  max_search_num = 3;
+  search_num = checker->size <= max_search_num ? checker->size
+                                                : max_search_num;
+
+  /* Check a few rotated positions. */
+  /* (this could be optimized by focusing on the min_atom_type) */
+  for (i_test = 0; i_test < search_num; i_test++) {
+
+    type_rot = checker->types_sorted[i_test];
+    mat_multiply_matrix_vector_id3(pos_rot, rot, checker->pos_sorted[i_test]);
+    for (k = 0; k < 3; k++) {
+      pos_rot[k] += test_trans[k];
+    }
+
+    /* Brute-force search for the rotated position. */
+    /* (this could be optimized by saving the sorted ValueWithIndex data */
+    /*  for the original Cell and using it to binary search for lower and */
+    /*  upper bounds on 'i'. For now though, brute force is good enough) */
```

Comment:
**Note:** The "this could be optimized" comment should be revised or removed if we switch to computing fractional-space distances because it assumes we are using cartesian distance.

Basically, it relies on the fact that, if two points are within a distance of SYMPREC to each other, then their values for the (non-squared) cartesian lattice point distance will also be within SYMPREC of each other.  This means that if the lattice point distance of `pos_rot` is `rot_value`, we only need to look at the region of values `rot_value - SYMPREC * FUZZ <= value <= rot_value + SYMPREC * FUZZ` (where `FUZZ` is something like `1 + 1e-7`)

(note: to prove the above property about cartesian lattice point distance, notice that the gradient always has a magnitude of 1 except at discontinuities, and that the value is still continuous even where the gradient is not)